### ISSUE TITLE
independent security audit

### ISSUE THREAD

**Issue by shibumi** [solokeys/solo#126]
I strongly suggest an independent security audit for the soft- and hardware.

**Comment by conorpp** [comment]
Nitrokey has plans to perform a security audit for the software at some point

**Comment by shibumi** [comment]
thanks for the answer and your transparency. What has nitrokey to do with this project?

**Comment by conorpp** [comment]
Nitrokey has plans to leverage some of our FIDO2 functionality and will contribute back.  One of their requirements is going through a security audit, so we both get to benefit :)

**Comment by shibumi** [comment]
@conorpp nice news!

**Comment by shibumi** [comment]
@0x0ece does this mean that solokey got an independend security audit?

**Comment by 0x0ece** [comment]
No audit and no funds to do one anytime soon.

Fo now FIDO2 L1 certification is the best assurance of security against online attacks.

Independently, we're trying to clean up issues starting from the ones with no activity. This morning we had 75+ issues, it's impossible to keep track of them. I feel issues should give a representation of what's happening and it's being worked on. For ideas and wishlist we should prob find a different place.

**Comment by MaPePeR** [comment]
> I feel issues should give a representation of what's happening and it's being worked on. For ideas and wishlist we should prob find a different place.

@0x0ece You could use Issue Labels and Milestones for this.

**Comment by shibumi** [comment]
Yes, there are labels and milestones.. I don't think there is a reason to close issues, especially issues asking for an independend security audit.. It looks really suspicious if you close such things without any word.

**Comment by 0x0ece** [comment]
This is a won’t fix, added label. Suspicious or not, there’s no budget.

**Comment by shibumi** [comment]
@0x0ece I don't think you need a budget for this.. I would just leave that issue open and maybe some nice security researcher out there would have an independend look on your hardware.I mean you could at least try to find a security researcher checking your device and if it's with a `We are unhackable`-campaign.. then all the researchers would come for free to just proof you wrong.

**Comment by wucke13** [comment]
I second this @shibumi .

**Comment by shibumi** [comment]
@0x0ece btw, does this all mean that you are not partnering up with nitrokeys? Because that's it sounds like for me.. @conorpp seemed pretty confident that there will be an independ security audit by nitrokey (they had a few in the past).

**Comment by jans23** [comment]
We (Nitrokey) did a short review of the most important security related aspects. For us, the firmware update was most crucial because it affects future firmware releases too and can't be updated itself. One shortcoming we identified and patched subsequently was that old firmware versions could be installed. Now with the patch, only newer firmware versions can be installed (downgrade protection). Because the review wasn't complete we didn't write a review. It would still be good to have a comprehensive security review. At a later stage, we may work more on this.

**Comment by 0x0ece** [comment]
I reopened then. But I confirm no current budget to do a security review.

**Comment by shibumi** [comment]
@jans23 Thanks a lot! This comment is gold for me. So, can we expect a review about this downgrade protection?

**Comment by 0x0ece** [comment]
Downgrade protection: https://github.com/solokeys/solo/pull/238

This may also be interesting, fault injection in USB: https://github.com/solokeys/solo/pull/224
(physical attack, not really realistic for 2FA, maybe for passwordless but unclear)

**Comment by jans23** [comment]
@shibumi We implemented the downgrade protection so somebody else should review it.

**Comment by 0x0ece** [comment]
Update. We run a security analysis with @Doyensec.

Blog post: https://solokeys.com/blogs/news/security-analysis-of-the-solo-firmware-by-doyensec

Full report: https://doyensec.com/resources/Doyensec_SoloKeys_TestingReport_Q12020_v3.pdf
